File: Nostos.Shared/Dtos/ReadingTrainingDtos.cs
using System;
using System.Collections.Generic;
using Nostos.Shared.Enums;

namespace Nostos.Shared.Dtos;

// --- STABLE RESPONSE ENVELOPE ---
// Every mutating command accepts a client id and idempotency key and returns
// this same semantic result, whether invoked through REST, MCP, or the UI.
public sealed record ReadingCommandResultDto(
    string Reply,
    object? Data,
    string StateVersion,
    bool Duplicate = false
);

public sealed record ReadingErrorDto(string Code);

// --- BASE REQUEST ---
public record ReadingCommandRequest(string ClientId, string IdempotencyKey);

// --- PRESCRIPTION / PLAN ---
public record ReadingPlanSessionRequest(
    string ClientId,
    string IdempotencyKey,
    Guid BookAssignmentId,
    ReadingMode Mode,
    int TargetMinutes,
    ReadingConstraint Constraint = ReadingConstraint.None
);

public record ReadingPrescriptionRequest(
    string ClientId,
    string IdempotencyKey,
    Guid? BookAssignmentId,
    ReadingMode? Mode,
    int? AvailableMinutes,
    ReadingConstraint Constraint = ReadingConstraint.None
);

// --- SESSION COMMANDS ---
public record ReadingStartSessionRequest(
    string ClientId,
    string IdempotencyKey,
    Guid? SessionId = null
);

public record ReadingStartNewSessionRequest(
    string ClientId,
    string IdempotencyKey,
    Guid? BookAssignmentId = null,
    ReadingMode? Mode = null
);

public record ReadingSessionCommandRequest(string ClientId, string IdempotencyKey);

public record ReadingCompleteSessionRequest(
    string ClientId,
    string IdempotencyKey,
    int? ReportedMinutes = null
);

public record ReadingRateSessionRequest(
    string ClientId,
    string IdempotencyKey,
    int Effort,
    int Focus,
    int? Rating = null
);

public record ReadingSkipRatingsRequest(string ClientId, string IdempotencyKey);

// --- BOOK ASSIGNMENT / QUEUE ---
public record ReadingAddBookAssignmentRequest(
    string ClientId,
    string IdempotencyKey,
    Guid BookId,
    ReadingMode Mode,
    bool MakeDefault = false
);

public record ReadingSetDefaultBookRequest(
    string ClientId,
    string IdempotencyKey,
    Guid BookAssignmentId,
    ReadingMode Mode
);

public record ReadingCompleteBookRequest(
    string ClientId,
    string IdempotencyKey,
    Guid BookAssignmentId
);

public record ReadingReorderQueueRequest(
    string ClientId,
    string IdempotencyKey,
    IReadOnlyList<Guid> AssignmentIds
);

// --- CAPTURE / INBOX ---
public record ReadingCaptureRequest(
    string ClientId,
    string IdempotencyKey,
    string Text,
    ReadingCaptureType Type,
    Guid? BookId = null,
    Guid? SessionId = null,
    string? ExternalId = null
);

public record ReadingResolveCaptureRequest(
    string ClientId,
    string IdempotencyKey,
    bool Keep,
    Guid? NoteId = null
);

public record ReadingPromoteCaptureRequest(
    string ClientId,
    string IdempotencyKey,
    Guid NoteId
);

// --- WEEKLY REVIEW ---
public record ReadingWeeklyReviewRequest(
    int Year,
    int Week
);

public record ReadingCommitWeeklyReviewRequest(
    string ClientId,
    string IdempotencyKey,
    int Year,
    int Week
);

// --- NOTIFICATION OUTBOX ---
public record ReadingAckNotificationRequest(string ClientId, string IdempotencyKey);

// --- GATEWAY (Hermes/Telegram connector) ---
public record ReadingGatewayDispatchRequest(string ClientId, string IdempotencyKey, string Text);

// --- IMPORT (Hermes data cutover) ---
public record ReadingImportDryRunRequest(
    string ClientId,
    string IdempotencyKey,
    IReadOnlyDictionary<string, string> Files
);

public record ReadingImportRequest(
    string ClientId,
    string IdempotencyKey,
    IReadOnlyDictionary<string, string> Files,
    bool CreateMissingBooks = false
);

// --- RESPONSE DTOs ---

public record ReadingTargetsDto(
    int EnduranceTargetMinutes,
    int DeepTargetMinutes,
    int RecoveryTargetMinutes,
    int EnduranceEstablishedMinutes,
    int DeepEstablishedMinutes,
    int RecoveryEstablishedMinutes
);

public record ReadingProgrammeDto(
    Guid Id,
    string TimezoneId,
    string StateVersion,
    ReadingTargetsDto Targets,
    bool DeloadActive
);

public record ReadingBookAssignmentDto(
    Guid Id,
    Guid BookId,
    string? BookTitle,
    string? BookAuthor,
    ReadingMode Mode,
    ReadingAssignmentStatus Status,
    int QueueOrder,
    bool IsDefault,
    DateTime CreatedAt,
    DateTime? StartedAt,
    DateTime? CompletedAt
);

public record ReadingSessionDto(
    Guid Id,
    Guid BookAssignmentId,
    Guid BookId,
    string? BookTitle,
    ReadingMode Mode,
    ReadingSessionStatus Status,
    int TargetMinutes,
    int PlannedTargetMinutes,
    ReadingConstraint Constraint,
    bool ProgressionEligible,
    bool CountsAsFailure,
    int AccumulatedSeconds,
    int MeasuredSeconds,
    int? ReportedMinutes,
    int Effort,
    int Focus,
    int? Rating,
    bool RatingsSkipped,
    DateTime PlannedAt,
    DateTime? StartedAt,
    DateTime? LastStartedAt,
    DateTime? PausedAt,
    DateTime? CompletedAt
);

public record ReadingWeekSummaryDto(
    string WeekKey,
    int CompletedSessions,
    int QualifyingSessions,
    int VolumeMinutes,
    int CompletionThreshold,
    bool ReviewCommitted
);

public record ReadingReviewDecisionDto(
    string WeekKey,
    ReadingMode Mode,
    int TargetBeforeMinutes,
    int TargetAfterMinutes,
    string DecisionKind,
    string Reason
);

// Semantic weekly review result shared by the read-only preview and the
// exact-once commit. Committed reviews expose the persisted StateVersionAfter;
// previews expose the current state version with Committed = false.
public record ReadingWeeklyReviewDto(
    string WeekKey,
    int IsoYear,
    int IsoWeek,
    bool Committed,
    DateTime? CommittedAt,
    int TotalVolumeMinutes,
    int PreviousWeekVolumeMinutes,
    IReadOnlyList<ReadingModeReviewDto> Modes,
    string StateVersion
);

public record ReadingModeReviewDto(
    ReadingMode Mode,
    int TargetBeforeMinutes,
    int TargetAfterMinutes,
    string DecisionKind,
    string Reason,
    int QualifyingCount,
    double CompletionRate,
    int? MedianEffort,
    int? MedianFocus,
    int NextConsecutiveIncreases
);

public record ReadingCaptureDto(
    Guid Id,
    string Text,
    ReadingCaptureType Type,
    Guid BookId,
    Guid? SessionId,
    string? ExternalId,
    bool Resolved,
    Guid? PromotedNoteId,
    DateTime CreatedAt
);

public record ReadingNotificationDto(
    Guid Id,
    string Kind,
    string PayloadJson,
    DateTime CreatedAt,
    DateTime? LeaseUntil,
    DateTime? AckedAt
);

public record ReadingImportReceiptDto(
    Guid Id,
    string SourceFingerprint,
    string ResultJson,
    DateTime CreatedAt
);

public record ReadingDashboardDto(
    ReadingProgrammeDto Programme,
    IReadOnlyList<ReadingBookAssignmentDto> Books,
    ReadingSessionDto? OpenSession,
    ReadingWeekSummaryDto? CurrentWeek
);

File: Nostos.Frontend/src/app/core/dtos/reading-training.dtos.ts
// Reading Training contracts for the Angular frontend.
//
// Mirrors Nostos.Shared/Dtos/ReadingTrainingDtos.cs + Nostos.Shared/Enums
// over the HTTP wire. ASP.NET Core's System.Text.Json emits camelCase property
// names, numeric enum values, and ISO-8601 strings for DateTime fields, so:
//   - every property below is camelCase (not the C# PascalCase name),
//   - enums are numeric literals matching the backend enum values,
//   - timestamps are `string` (non-nullable) or `string | null` (nullable),
//   - GUIDs are `string`.
//
// The stable command envelope `ReadingCommandResult<T>` wraps every mutating
// command response (`reply`, `data`, `stateVersion`, `duplicate`); the
// notification lease/ack endpoints are NOT command envelopes and use their own
// response shapes below.

// --- ENUMS (numeric, matches Nostos.Shared/Enums) ---

export enum ReadingMode {
  Endurance = 0,
  Deep = 1,
  Recovery = 2,
}

export enum ReadingConstraint {
  None = 0,
  TimeConstrained = 1,
  FatigueConstrained = 2,
}

export enum ReadingAssignmentStatus {
  Active = 0,
  Queued = 1,
  Completed = 2,
  Archived = 3,
}

export enum ReadingSessionStatus {
  Idle = 0,
  Planned = 1,
  Active = 2,
  Paused = 3,
  AwaitingFeedback = 4,
  Completed = 5,
  Cancelled = 6,
}

export enum ReadingCaptureType {
  Thought = 0,
  Question = 1,
  Bookmark = 2,
}

// --- STABLE RESPONSE ENVELOPE ---
// Every mutating command returns this semantic result, whether invoked
// through REST, MCP, or the UI. `data` is null when the command has no
// payload (e.g. status with no active session).
export interface ReadingCommandResult<T> {
  reply: string;
  data: T | null;
  stateVersion: string;
  duplicate: boolean;
}

export interface ReadingError {
  code: string;
}

/**
 * Narrowing guard for the semantic error payload the backend places in a
 * command envelope's `data` slot (`{ code }`, e.g. `not_initialized`). No
 * successful response DTO carries a `code` field, so any non-null payload
 * with a non-empty string `code` is a semantic error — even inside an
 * HTTP-200 envelope, which the transport layer cannot otherwise distinguish
 * from a successful payload.
 */
export function isReadingError(value: unknown): value is ReadingError {
  return (
    typeof value === 'object' &&
    value !== null &&
    typeof (value as { code?: unknown }).code === 'string' &&
    (value as { code?: unknown }).code !== ''
  );
}

// --- BASE REQUEST ---
export interface ReadingCommandRequest {
  clientId: string;
  idempotencyKey: string;
}

// Sessions' pause/resume/cancel take exactly the base command fields.
export type ReadingSessionCommandRequest = ReadingCommandRequest;

// --- SESSION COMMANDS ---
export interface ReadingPlanSessionRequest extends ReadingCommandRequest {
  bookAssignmentId: string;
  mode: ReadingMode;
  targetMinutes: number;
  constraint?: ReadingConstraint;
}

export interface ReadingStartSessionRequest extends ReadingCommandRequest {
  sessionId?: string;
}

export interface ReadingStartNewSessionRequest extends ReadingCommandRequest {
  bookAssignmentId?: string;
  mode?: ReadingMode;
}

export interface ReadingCompleteSessionRequest extends ReadingCommandRequest {
  reportedMinutes?: number;
}

export interface ReadingRateSessionRequest extends ReadingCommandRequest {
  effort: number;
  focus: number;
  rating?: number;
}

export interface ReadingSkipRatingsRequest extends ReadingCommandRequest {}

// --- BOOK ASSIGNMENT / QUEUE ---
export interface ReadingAddBookAssignmentRequest extends ReadingCommandRequest {
  bookId: string;
  mode: ReadingMode;
  makeDefault?: boolean;
}

export interface ReadingSetDefaultBookRequest extends ReadingCommandRequest {
  bookAssignmentId: string;
  mode: ReadingMode;
}

export interface ReadingCompleteBookRequest extends ReadingCommandRequest {
  bookAssignmentId: string;
}

export interface ReadingReorderQueueRequest extends ReadingCommandRequest {
  assignmentIds: string[];
}

/** PATCH /api/reading/books/{assignmentId}/mode — the assignment id lives in the URL. */
export interface ReadingChangeBookModeRequest extends ReadingCommandRequest {
  mode: ReadingMode;
}

/** DELETE /api/reading/books/{assignmentId} — the assignment id lives in the URL. */
export interface ReadingRemoveBookAssignmentRequest extends ReadingCommandRequest {}

/** Server data for a successful mode change (absorbed collider optional). */
export interface ReadingChangeBookModeData {
  assignmentId: string;
  bookId: string;
  previousMode: ReadingMode;
  mode: ReadingMode;
  queueOrder: number;
  /** The mode's default-slot sentinel when the assignment becomes the mode default, else null. */
  defaultSlot: string | null;
  collisionAbsorbed: boolean;
  absorbedAssignmentId: string | null;
}

/** Server data for a successful queue removal. */
export interface ReadingRemoveBookAssignmentData {
  assignmentId: string;
  bookId: string;
  mode: ReadingMode;
  queueOrder: number;
}

// --- CAPTURE / INBOX ---
export interface ReadingCaptureRequest extends ReadingCommandRequest {
  text: string;
  type: ReadingCaptureType;
  bookId?: string;
  sessionId?: string;
  externalId?: string;
}

export interface ReadingResolveCaptureRequest extends ReadingCommandRequest {
  keep: boolean;
  noteId?: string;
}

export interface ReadingPromoteCaptureRequest extends ReadingCommandRequest {
  noteId: string;
}

// --- WEEKLY REVIEW ---
export interface ReadingCommitWeeklyReviewRequest extends ReadingCommandRequest {
  year: number;
  week: number;
}

// --- RESPONSE DTOs ---

export interface ReadingTargets {
  enduranceTargetMinutes: number;
  deepTargetMinutes: number;
  recoveryTargetMinutes: number;
  enduranceEstablishedMinutes: number;
  deepEstablishedMinutes: number;
  recoveryEstablishedMinutes: number;
}

export interface ReadingProgramme {
  id: string;
  timezoneId: string;
  stateVersion: string;
  targets: ReadingTargets;
  deloadActive: boolean;
}

export interface ReadingBookAssignment {
  id: string;
  bookId: string;
  bookTitle: string | null;
  bookAuthor: string | null;
  mode: ReadingMode;
  status: ReadingAssignmentStatus;
  queueOrder: number;
  isDefault: boolean;
  createdAt: string;
  startedAt: string | null;
  completedAt: string | null;
}

export interface ReadingSession {
  id: string;
  bookAssignmentId: string;
  bookId: string;
  bookTitle: string | null;
  mode: ReadingMode;
  status: ReadingSessionStatus;
  targetMinutes: number;
  plannedTargetMinutes: number;
  constraint: ReadingConstraint;
  progressionEligible: boolean;
  countsAsFailure: boolean;
  accumulatedSeconds: number;
  measuredSeconds: number;
  reportedMinutes: number | null;
  effort: number;
  focus: number;
  rating: number | null;
  ratingsSkipped: boolean;
  plannedAt: string;
  startedAt: string | null;
  lastStartedAt: string | null;
  pausedAt: string | null;
  completedAt: string | null;
}

export interface ReadingWeekSummary {
  weekKey: string;
  completedSessions: number;
  qualifyingSessions: number;
  volumeMinutes: number;
  completionThreshold: number;
  reviewCommitted: boolean;
}

export interface ReadingModeReview {
  mode: ReadingMode;
  targetBeforeMinutes: number;
  targetAfterMinutes: number;
  // Decision kind is a backend string, not an enum: lowercase
  // "increase" | "hold" | "deload" | "consolidate".
  decisionKind: string;
  reason: string;
  qualifyingCount: number;
  completionRate: number;
  medianEffort: number | null;
  medianFocus: number | null;
  nextConsecutiveIncreases: number;
}

export interface ReadingWeeklyReview {
  weekKey: string;
  isoYear: number;
  isoWeek: number;
  committed: boolean;
  committedAt: string | null;
  totalVolumeMinutes: number;
  previousWeekVolumeMinutes: number;
  modes: ReadingModeReview[];
  stateVersion: string;
}

export interface ReadingCapture {
  id: string;
  text: string;
  type: ReadingCaptureType;
  bookId: string;
  sessionId: string | null;
  externalId: string | null;
  resolved: boolean;
  promotedNoteId: string | null;
  createdAt: string;
}

export interface ReadingDashboard {
  programme: ReadingProgramme;
  books: ReadingBookAssignment[];
  openSession: ReadingSession | null;
  currentWeek: ReadingWeekSummary | null;
}

// --- NOTIFICATION OUTBOX (lease/ack are NOT command envelopes) ---

export interface TargetReachedNotificationPayload {
  notificationId: string;
  sessionId: string;
  bookId: string;
  mode: ReadingMode;
  plannedTargetMinutes: number;
  effectiveElapsedSeconds: number;
  message: string;
}

// One row handed to a delivery worker: the typed payload plus the lease.
export interface ReadingNotification {
  notificationId: string;
  payload: TargetReachedNotificationPayload;
  leaseUntil: string;
}

// Idempotent ack response: 200 with `{ notificationId, acknowledged: true }`.
export interface ReadingAckNotificationResult {
  notificationId: string;
  acknowledged: boolean;
}
